### PR TITLE
opt: Increase performance of small queries

### PR DESCRIPTION
**Pull request by andy-kimball**
This PR contains two commits that increase the performance of
small, simple queries:

- Improve statistics building perf
- Streamline join name dup detection

Together, these changes result in these bench test improvements:
```
name                      old time/op  new time/op  delta
Phases/kv-read/Explore    28.0µs ± 2%  26.1µs ± 3%  -6.94%  (p=0.008 n=5+5)
Phases/planning1/Explore  10.5µs ± 3%  10.3µs ± 2%    ~     (p=0.310 n=5+5)
Phases/planning2/Explore  31.2µs ± 2%  30.7µs ± 3%    ~     (p=0.222 n=5+5)
Phases/planning3/Explore  36.2µs ± 2%  34.3µs ± 3%  -5.28%  (p=0.008 n=5+5)
Phases/planning4/Explore  44.2µs ± 1%  41.5µs ± 2%  -6.00%  (p=0.008 n=5+5)
Phases/planning5/Explore  35.9µs ± 1%  34.1µs ± 2%  -5.10%  (p=0.008 n=5+5)
Phases/planning6/Explore  63.8µs ± 3%  59.3µs ± 3%  -7.12%  (p=0.008 n=5+5)
```

### CODE REVIEW

File: pkg/sql/opt/memo/statistics_builder.go [hunk 1]
```diff
@@ -710,37 +662,24 @@ func (sb *statisticsBuilder) buildJoin(ev ExprView, relProps *props.Relational)
...
-		colsToRemove.DifferenceWith(leftProps.OutputCols)
+		s.ColStats.RemoveIntersecting(rightProps.OutputCols)
 
 	case opt.FullJoinOp, opt.FullJoinApplyOp:
 		// Do not keep any column stats.
-
-	case opt.InnerJoinOp, opt.InnerJoinApplyOp:
-		// Keep all column stats.
-		colsToRemove = opt.ColSet{}
+		s.ColStats.Clear()
```
Comment (RaduBerinde) [inline pkg/sql/opt/memo/statistics_builder.go]: I think this belongs in the case above

File: pkg/sql/opt/props/col_stats_map.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+)
+
+const (
+	// initialColStatsCap is the initial number of column statistics that can be
+	// stored in a ColStatsMap without triggering allocations.
+	initialColStatsCap = 3
+)
+
+type colStatKey struct {
+	prefix uint32
```
Comment (RaduBerinde) [inline pkg/sql/opt/props/col_stats_map.go]: A "type prefixID uint32" would make the code a bit cleaner.

**Review by RaduBerinde — APPROVED** [review]
Very cool!
LGTM except one comment in the stats builder.

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29026)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/statistics_builder.go, line 680 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29026#-LKdFOJG6pLSmRaDqmgH-r2-680:-LKjU-AK-IH_R9di27ay:b-c39n69) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e043a717b8fef754fde194c693b51f9f3e808/pkg/sql/opt/memo/statistics_builder.go#L680)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think this belongs in the case above
</blockquote></details>

I don't follow, what do you mean?
___
*[pkg/sql/opt/props/col_stats_map.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29026#-LKdFOJG6pLSmRaDqmgK-r2-28:-LKw0wjZ-w-ytEp6zxwf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e043a717b8fef754fde194c693b51f9f3e808/pkg/sql/opt/props/col_stats_map.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

A "type prefixID uint32" would make the code a bit cleaner.
</blockquote></details>

Done.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29026#-:-LKw5UFSDF6R9XoBko-L:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/memo/statistics_builder.go, line 680 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29026#-LKdFOJG6pLSmRaDqmgH-r2-680:-LKw4gieCOehvhl2sz8y:b-ljkrgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e043a717b8fef754fde194c693b51f9f3e808/pkg/sql/opt/memo/statistics_builder.go#L680)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't follow, what do you mean?
</blockquote></details>

Never mind, I misread the diff on my phone and thought this was still under the `InnerJoinOp` case.

**Pull request merged**